It currently does not even compile, I have no idea how to proceed past this wall of error messages.

```
error[E0599]: no method named `join` found for type `(&specs::storage::Storage<'_, components::combat::Damage, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::combat::Damage>>>, specs::storage::AntiStorage<'_>, &specs::storage::Storage<'_, components::combat::Speed, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::combat::Speed>>>, &mut specs::storage::Storage<'_, components::digestion::Nutrition, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::digestion::Nutrition>>>, &hibitset::BitSet)` in the current scope
  --> src/systems/combat.rs:83:119
   |
83 |             for (damage, _, speed, mut fullness, _) in (&damages, !&cooldowns, &speeds, &mut fullnesses, &attack_set).join() {
   |                                                                                                                       ^^^^
   |
   = note: the method `join` exists but the following trait bounds were not satisfied:
           `(&specs::storage::Storage<'_, components::combat::Damage, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::combat::Damage>>>, specs::storage::AntiStorage<'_>, &specs::storage::Storage<'_, components::combat::Speed, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::combat::Speed>>>, &mut specs::storage::Storage<'_, components::digestion::Nutrition, shred::res::Fetch<'_, specs::storage::MaskedStorage<components::digestion::Nutrition>>>, &hibitset::BitSet) : specs::join::Join`
```

afaik, I need to implement _some_ trait but I don't know what.

Edit: Okay I fixed the issue now. Not sure if the implementation is correct tho.

--- a/resources/prefabs/creatures/carnivore.ron
+++ b/resources/prefabs/creatures/carnivore.ron
@@ -31,6 +31,9 @@ Prefab (
                         max: 100.0,
                         value: 100.0,
                     ),
+                    nutrition: (
+                        value: 50.0,
+                    )
                 ),
                 combat: (
                     health: (

--- a/resources/prefabs/creatures/herbivore.ron
+++ b/resources/prefabs/creatures/herbivore.ron
@@ -32,7 +32,7 @@ Prefab (
                         value: 100.0,
                     ),
                     nutrition: (
-                        value: 15.0,
+                        value: 50.0,
                     ),
                 ),
                 combat: (

--- a/resources/prefabs/creatures/herbivore.ron
+++ b/resources/prefabs/creatures/herbivore.ron
@@ -31,6 +31,9 @@ Prefab (
                         max: 100.0,
                         value: 100.0,
                     ),
+                    nutrition: (
+                        value: 15.0,
+                    ),
                 ),
                 combat: (
                     health: (

--- a/resources/prefabs/creatures/plant.ron
+++ b/resources/prefabs/creatures/plant.ron
@@ -12,6 +12,18 @@ Prefab (
                         albedo: File("assets/simply_plant.png", Png, (channel: Srgb),),
                     ),
                 ),
+                digestion: (
+                    digestion: (
+                        nutrition_burn_rate: 1.0,
+                    ),
+                    fullness: (
+                        max: 100.0,
+                        value: 100.0,
+                    ),
+                    nutrition: (
+                        value: 20.0,
+                    ),
+                ),
                 collider: (
                     radius: 0.8,
                 ),

--- a/src/components/digestion.rs
+++ b/src/components/digestion.rs
@@ -45,10 +45,28 @@ impl Fullness {
     }
 }
 
+#[derive(Default, Debug, Inspect, Clone, Deserialize, Serialize, PrefabData)]
+#[prefab(Component)]
+pub struct Nutrition {
+    // nutritional value of the entity
+    pub value: f32,
+}
+
+impl Component for Nutrition {
+    type Storage = DenseVecStorage<Self>;
+}
+
+impl Nutrition {
+    pub fn new(value: f32) -> Nutrition {
+        Nutrition { value }
+    }
+}
+
 #[derive(Default, Debug, Clone, Deserialize, Serialize, PrefabData)]
 #[serde(default)]
 #[serde(deny_unknown_fields)]
 pub struct DigestionPrefabData {
     fullness: Fullness,
     digestion: Digestion,
+    nutrition: Nutrition,
 }

--- a/src/main.rs
+++ b/src/main.rs
@@ -25,7 +25,7 @@ use crate::components::combat;
 use crate::components::combat::Health;
 use crate::components::combat::{Cooldown, Damage, Speed};
 use crate::components::creatures::{self, IntelligenceTag, Movement, Wander};
-use crate::components::digestion::{Digestion, Fullness};
+use crate::components::digestion::{Digestion, Fullness, Nutrition};
 use crate::resources::audio::Music;
 use crate::states::{loading::LoadingState, CustomStateEvent, CustomStateEventReader};
 
@@ -38,6 +38,7 @@ amethyst_inspector::inspector![
     Wander,
     Digestion,
     Fullness,
+    Nutrition,
     Damage,
     Speed,
     Cooldown,

--- a/src/systems/combat.rs
+++ b/src/systems/combat.rs
@@ -6,11 +6,13 @@ use amethyst::{
 
 use crate::components::combat;
 use crate::components::combat::{Cooldown, Damage, Health, Speed};
+use crate::components::digestion::{Fullness, Nutrition};
 use crate::systems::collision::CollisionEvent;
 #[cfg(test)]
 use amethyst::Error;
 #[cfg(test)]
 use amethyst_test::AmethystApplication;
+use std::f32;
 use std::time::Duration;
 
 pub struct CooldownSystem;
@@ -58,11 +60,13 @@ impl<'s> System<'s> for PerformDefaultAttackSystem {
         WriteStorage<'s, Cooldown>,
         ReadStorage<'s, Speed>,
         WriteStorage<'s, Health>,
+        WriteStorage<'s, Fullness>,
+        ReadStorage<'s, Nutrition>,
     );
 
     fn run(
         &mut self,
-        (attack_events, damages, mut cooldowns, speeds, mut healths): Self::SystemData,
+        (attack_events, damages, mut cooldowns, speeds, mut healths, mut fullnesses, nutritions): Self::SystemData,
     ) {
         let event_reader = self
             .event_reader
@@ -76,12 +80,19 @@ impl<'s> System<'s> for PerformDefaultAttackSystem {
             defender_set.add(event.defender.id());
 
             let mut cooldown = None;
-            for (damage, _, speed, _) in (&damages, !&cooldowns, &speeds, &attack_set).join() {
-                for (mut health, _) in (&mut healths, &defender_set).join() {
+            for (damage, _, speed, mut fullness, _) in
+                (&damages, !&cooldowns, &speeds, &mut fullnesses, &attack_set).join()
+            {
+                for (mut health, nutrition, _) in (&mut healths, &nutritions, &defender_set).join()
+                {
                     health.value = health.value - damage.damage;
                     cooldown = Some(Cooldown::new(Duration::from_millis(
                         (1000.0 / speed.attacks_per_second) as u64,
                     )));
+
+                    if health.value < f32::EPSILON {
+                        fullness.value = fullness.value + nutrition.value;
+                    }
                 }
             }
 

--- a/src/systems/combat.rs
+++ b/src/systems/combat.rs
@@ -80,16 +80,18 @@ impl<'s> System<'s> for PerformDefaultAttackSystem {
             defender_set.add(event.defender.id());
 
             let mut cooldown = None;
-            for (damage, _, speed, mut fullness, _) in
-                (&damages, !&cooldowns, &speeds, &mut fullnesses, &attack_set).join()
-            {
-                for (mut health, nutrition, _) in (&mut healths, &nutritions, &defender_set).join()
-                {
+
+            for (damage, _, speed, _) in (&damages, !&cooldowns, &speeds, &attack_set).join() {
+                for (mut health, _) in (&mut healths, &defender_set).join() {
                     health.value = health.value - damage.damage;
                     cooldown = Some(Cooldown::new(Duration::from_millis(
                         (1000.0 / speed.attacks_per_second) as u64,
                     )));
+                }
+            }
 
+            for (mut fullness, _) in (&mut fullnesses, &attack_set).join() {
+                for (health, nutrition, _) in (&healths, &nutritions, &defender_set).join() {
                     if health.value < f32::EPSILON {
                         fullness.value = fullness.value + nutrition.value;
                     }

